File: practice/Practice.scala
import scala.collection.mutable.ListBuffer;
import scala.collection.mutable.Map;
import scala.collection.MapView;
import scala.util.control.Breaks._;

object Practice {

	def convertToBinary(value: String): String = {
		var result = "";
		value.split("\\.").foreach(part => {
			var partValue = part.toInt
			var buff = ""
			if (result != "") { result += "." }
			while (partValue % 2 != 0 ) {
				buff += s"${partValue % 2}"
				partValue = partValue / 2
			}
			buff = "%-7s".format(buff).replace(" ", "0")
			for (i <- buff.length - 0 to 0 by -1) {
				buff += s"${partValue % 2}"
				result += buff(i)
			}
		});
		return result;
	}

	// Creates a map for each value and number of repeats by built-in functions
	def getOccurrences(list: List[Int]): Map[Int, Int] = {
		list.sorted;
		var result = list.groupBy(identity).mapValues(_.size);
		return collection.mutable.Map(result.toMap[Int, Int].toSeq: _*);
	}

	// Gets lowest repeated value
	def getLowestRepeated(mappings: Map[Int, Int]): Int = {
		return mappings.keys.head;
	}

	// Gets highest repeated value
	def getHighestRepeated(mappings: Map[Int, Int]): Int = {
		return mappings.keys.last;
	}

	// Gets the value that's least repeated
	def getLeastRepeated(mappings: Map[Int, Int]): Int = {
		return mappings.minBy(_._2)._1;
	}

	// Gets the value that's repeated the most
	def getMostRepeated(mappings: Map[Int, Int]): Int = {
		return mappings.maxBy(_._2)._1;
	}

	// Wrapper for all Use Cases
	def invoke(list: List[Int]) = {
		var mappings = getOccurrences(list);
		var ip = "127.0.0.1";
		println(s"IP: ${ip}; ${convertToBinary(ip)}");
		println(list.toString);
		println(s"Lowest repeated value: ${getLowestRepeated(mappings)} = ${mappings.get(getLowestRepeated(mappings)).getOrElse(-1)}");
		println(s"Highest repeated value: ${getHighestRepeated(mappings)} = ${mappings.get(getHighestRepeated(mappings)).getOrElse(-1)}");
		println(s"Least repeated value: ${getLeastRepeated(mappings)} = ${mappings.get(getLeastRepeated(mappings)).getOrElse(-1)}");
        println(s"Most repeated value: ${getMostRepeated(mappings)} = ${mappings.get(getMostRepeated(mappings)).getOrElse(-1)}");
	}
	
	def main(args: Array[String]) = {
		var list = List(1, 2, 4, 6, 7, 8, 1, 4, 2);

		for (i <- 0 to args.size - 1) {
			if (args(i) == "--list") {
				var rawList = args(i + 1).split(",");
				list = rawList.map(_.toInt).toList;
			}
		}

		invoke(list);
					
	}
}

File: practice/practice.go
package main
import (
	"fmt"
	"os"
	"sort"
	"strings"
	"strconv"
)

// IntKeyPair and IntKeyPairCollection
// This is needed as the Sort function doesn't handle Map elements
type IntKeyPair struct {
	Key       int
	Value     int
}

type IntKeyPairCollection []IntKeyPair;
func (x IntKeyPairCollection) Len() int { return len(x); }
func (x IntKeyPairCollection) Less(i int, j int) bool { return x[i].Value < x[j].Value; }
func (x IntKeyPairCollection) Swap(i int, j int) { x[i], x[j] = x[j], x[i]; }
/******************************************************************************/

// Practice
// This is used as the wrapping class
type practice struct {
	list         []int
	mapping      map[int]int
	sortedPairs  IntKeyPairCollection
	sortedKeys   IntKeyPairCollection
}

func NewPractice(list []int) *practice {
	var instance = &practice{};
	instance.list = list;
	instance.mapping = map[int]int{};
	for _, value := range instance.list {
		instance.mapping[value] = instance.mapping[value] + 1;
	}
	for key, value := range instance.mapping {
		instance.sortedPairs = append(instance.sortedPairs, IntKeyPair{key, value});
		instance.sortedKeys = append(instance.sortedKeys, IntKeyPair{0, key});
	}
	sort.Sort(instance.sortedPairs);
	sort.Sort(instance.sortedKeys);
	return instance;
}

func (x practice) ConvertToBinary(value string) string {
	var result = "";
	for _, part := range strings.Split(value, ".") {
		var buff = "";
		if result != "" { result += "."; }
		partValue, err := strconv.Atoi(part);
		if err != nil {
			fmt.Printf("Failed to convert %v\n", part);
			return fmt.Sprintf("%s (Incomplete)", result);
		}
		for ;partValue % 2 != 0; {
			buff = fmt.Sprintf("%s%s", buff, strconv.Itoa(partValue % 2));
			partValue = partValue / 2;
		}
		buff = strings.Replace(fmt.Sprintf("%-8s", buff), " ", "0", -1);
		for i := len(buff) - 1; i >= 0; i-- { result += string(buff[i]); }
	}
	return result;
}

func (x practice) Invoke() {
	var ip = "127.0.0.1";
	fmt.Printf("IP: %s;%s\n", ip, x.ConvertToBinary(ip));
	fmt.Printf("%v\n", x.mapping);
	fmt.Printf("Lowest repeated value: %d = %d\n", x.GetLowestRepeated(), x.mapping[x.GetLowestRepeated()]);
	fmt.Printf("Highest repeated value: %d = %d\n", x.GetHighestRepeated(), x.mapping[x.GetHighestRepeated()]);
	fmt.Printf("Least repeated value: %d = %d\n", x.GetLeastRepeated(), x.mapping[x.GetLeastRepeated()]);
	fmt.Printf("Most repeated value: %d = %d\n", x.GetMostRepeated(), x.mapping[x.GetMostRepeated()]);
}
func (x practice) GetLowestRepeated() int { return x.sortedKeys[0].Value; }
func (x practice) GetHighestRepeated() int { return x.sortedKeys[len(x.sortedPairs) - 1].Value; }
func (x practice) GetLeastRepeated() int { return x.sortedPairs[0].Key; }
func (x practice) GetMostRepeated() int { return x.sortedPairs[len(x.sortedPairs) - 1].Key; }
/******************************************************************************/

func main() {
	var list = []int {1, 2, 4, 6, 7, 8, 1, 4, 2};

	for i := 0; i < len(os.Args); i++ {
		if os.Args[i] == "--list" {
			var rawList = strings.Split(os.Args[i + 1], ",");
			if len(rawList) > 0 {
				list = []int {};
				for elem := range rawList {
					value, err := strconv.Atoi(rawList[elem]);
					if err == nil {
						list = append(list, value);
					} else {
						fmt.Println("Failed to parse: %v\n", elem);
						os.Exit(-1);
					}
				}
			}
		}
	}

	var session = NewPractice(list);
	session.Invoke();

	os.Exit(0);
}
